Sources/BrowserServicesKit/ContentBlocking/ContentBlockingRulesLookupTask.swift

```diff
@@ -28,31 +28,45 @@ extension ContentBlockerRulesManager {
 
         private let sourceManagers: [ContentBlockerRulesSourceManager]
 
-        public private(set) var result: [LookupResult]?
-
         init(sourceManagers: [ContentBlockerRulesSourceManager]) {
             self.sourceManagers = sourceManagers
         }
 
-        func lookupCachedRulesLists() async throws {
+        func lookupCachedRulesLists() throws -> [LookupResult]  {
+
+            let models = sourceManagers.compactMap { $0.makeModel() }
+            if models.count != sourceManagers.count {
+                // We should only load rule lists, in case we can match every one of the expected ones
+                throw WKError(.contentRuleListStoreLookUpFailed)
+            }
 
             var result = [LookupResult]()
-            for sourceManager in sourceManagers {
-                guard let model = sourceManager.makeModel() else {
-                    throw WKError(.contentRuleListStoreLookUpFailed)
-                }
+            let group = DispatchGroup()
+
+            for model in models {
+                group.enter()
 
-                guard let ruleList = try await Task(operation: { @MainActor in
-                    try await WKContentRuleListStore.default().contentRuleList(forIdentifier: model.rulesIdentifier.stringValue)
-                }).value else {
-                    // All lists must be found for this to be considered successful
-                    throw WKError(.contentRuleListStoreLookUpFailed)
+                DispatchQueue.main.async {
+                    // This needs to be called from the main thread.
+                    WKContentRuleListStore.default().lookUpContentRuleList(forIdentifier: model.rulesIdentifier.stringValue) { ruleList, error in
+                        guard let ruleList, error == nil else {
+                            group.leave()
+                            return
+                        }
+
+                        result.append((ruleList, model))
+                        group.leave()
+                    }
                 }
+            }
+
+            let operationResult = group.wait(timeout: .now() + 6)
 
-                result.append((ruleList, model))
+            guard operationResult == .success, result.count == models.count else {
+                throw WKError(.contentRuleListStoreLookUpFailed)
```

Comment:
Do we currently get enough signal that this issue is happening?

I'd want to know how often this specific case happens to users as it provides some interesting information about the `lookUpContentRuleList` API, though to the caller we don't differentiate between reasons why this function failed. From what I see, a timeout is indistinguishable from a failed `models.count != sourceManagers.count` check above, for example.

Reply:
I think this should not happen, unless WebKit has latency, which is not even actionable by us.

We actually have different signal added recently by Shilpa: startup distribution of lookup vs last fetched vs recompile flows. I'm hesitant to ad anything here without good reason, and it's not like timeout affects app state in some really bad way - will just require us to recompile (which does a lookup anyway).

The checks for models are onl for the app upgrades really, when we change the set of Content Blocking data sets (e.g. when introducing CTL or Attribution features).

That's why I don't really think we need the Pixels here, what do you think?

Reply:
> We actually have different signal added recently by Shilpa: startup distribution of lookup vs last fetched vs recompile flows

Great point! As long as we have a way to get a hunch when this case could be happening (and I agree it shouldn't), I'm happy here. Thanks!